File: MultiLineSearch/DteInitializer.vb
Imports Microsoft.VisualStudio
Imports Microsoft.VisualStudio.Shell.Interop


Friend Delegate Sub Action()


''' <summary>
''' Initalizes the DTE object.
''' </summary>
''' <remarks></remarks>
Friend Class DteInitializer
    Implements IVsShellPropertyEvents
    Private ReadOnly shellService As IVsShell
    Private cookie As UInteger
    Private ReadOnly callback As Action


    Friend Sub New(ByVal shellService As IVsShell, ByVal callback As Action)
        Dim hr As Integer

        Me.shellService = shellService
        Me.callback = callback

        ' Set an event handler to detect when the IDE is fully initialized
        hr = Me.shellService.AdviseShellPropertyChanges(Me, Me.cookie)

        Microsoft.VisualStudio.ErrorHandler.ThrowOnFailure(hr)
    End Sub


    Private Function IVsShellPropertyEvents_OnShellPropertyChange(ByVal propid As Integer, ByVal var As Object) As Integer Implements IVsShellPropertyEvents.OnShellPropertyChange
        Dim hr As Integer
        Dim isZombie As Boolean

        If propid = CInt(__VSSPROPID.VSSPROPID_Zombie) Then
            isZombie = CBool(var)

            If Not isZombie Then
                ' Release the event handler to detect when the IDE is fully initialized
                hr = Me.shellService.UnadviseShellPropertyChanges(Me.cookie)
                Microsoft.VisualStudio.ErrorHandler.ThrowOnFailure(hr)
                Me.cookie = 0
                Me.callback()
            End If
        End If
        Return VSConstants.S_OK
    End Function
End Class

File: MultiLineSearch.Shared/DteInitializer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Microsoft.VisualStudio;
using Microsoft.VisualStudio.Shell.Interop;
using Microsoft.VisualStudio.Shell;


namespace Helixoft.MultiLineSearch
{


    /// <summary>
    /// Initializes the DTE object.
    /// </summary>
    /// <remarks></remarks>
    internal class DteInitializer : IVsShellPropertyEvents
    {
        private readonly IVsShell shellService;
        private uint cookie;

        private readonly Action callback;

        internal DteInitializer(IVsShell shellService, Action callback)
        {
            ThreadHelper.ThrowIfNotOnUIThread();

            int hr = 0;

            this.shellService = shellService;
            this.callback = callback;

            // Set an event handler to detect when the IDE is fully initialized
            hr = this.shellService.AdviseShellPropertyChanges(this, out this.cookie);

            Microsoft.VisualStudio.ErrorHandler.ThrowOnFailure(hr);
        }


        private int IVsShellPropertyEvents_OnShellPropertyChange(int propid, object var)
        {
            ThreadHelper.ThrowIfNotOnUIThread();

            int hr = 0;
            bool isZombie = false;

            if (propid == Convert.ToInt32(__VSSPROPID.VSSPROPID_Zombie))
            {
                isZombie = Convert.ToBoolean(var);

                if (!isZombie)
                {
                    // Release the event handler to detect when the IDE is fully initialized
                    hr = this.shellService.UnadviseShellPropertyChanges(this.cookie);
                    Microsoft.VisualStudio.ErrorHandler.ThrowOnFailure(hr);
                    this.cookie = 0;
                    this.callback();
                }
            }
            return VSConstants.S_OK;
        }
        int IVsShellPropertyEvents.OnShellPropertyChange(int propid, object var)
        {
            ThreadHelper.ThrowIfNotOnUIThread();

            return IVsShellPropertyEvents_OnShellPropertyChange(propid, var);
        }
    }

}